Create and expose php_sys_symlink() and php_sys_link()

cmb69 opened the pull request:
On POSIX compliant systems these are just macros which map directly to
symlink() and link(), respectively.  On Windows they are implemented as
functions which try to conform as closely as possible to the POSIX
specification with regard to PHP's needs.  Actually, they have been
inlined in link_win32.c to implement PHP's symlink() and link()
functions.  It is noteworthy to mention that POSIX systems store the
error in `errno`, while on Windows `GetLastError()` has to be used to
retrieve the error (analogous to php_sys_readlink()).

Future scope: merge link.c and link_win32.c

cmb69 commented:
Further explanation: the basic desire is to merge link.c and link_win32.c to have a single link.c for all platforms, to avoid issues similar to https://bugs.php.net/76459. Since link_win32.c has a [TODO](https://github.com/php/php-src/blob/php-7.2.8/ext/standard/link_win32.c#L46-L50) which suggests to first implement `php_link` and `php_symlink` (which should actually be `php_sys_link` and `php_sys_symlink` to conform to `php_sys_readlink`), I've done this in this PR.

@weltling Since this mostly concerns Windows, what do you think? Are `php_sys_link` and `php_sys_symlink` ready for general use? Otherwise an alternative might be to not expose them, but rather to declare them static inside of link(_win32).c.

weltling commented:
@cmb69 thanks for this work. The code in `link_win32.c` is in fact shipped in production binaries, so it's as stable as that.

Regarding the approach there are several notes from my side. I'd suggest to move this into `win32/ioutil.*` space. It is a platform specific detail and it would make the code in `zend_virtual_cwd.*` more slim.

Also, the ioutil stuff exposes the wide variants separately, so IMO it would make sense to rewrite for that, too. As an example, there's `php_win32_ioutil_open` and `php_win32_ioutil_open_w`. The goal of this is to avoid unnecessary wide<>char conversions in situations, where one `php_win32_iouti` function is called inside of another.

In further, the POSIX compatibility of these functions can be improved by the usage of `SET_ERRNO_FROM_WIN32_CODE` as in accordance with the corresponding man pages.

There are yet some places in the core, that need such refactoring. I'd suggest these new functions to go into win32. At some point many ifdefs can be avoided even more.

Thanks.

cmb69 commented:
@weltling Thanks (especially for hinting at `SET_ERRNO_FROM_WIN32_CODE`, which I thought was missing)! I have addressed all your suggestions (except for refactoring other functions, which I'll leave as TODO for the future).

weltling commented:
@cmb69 nice :) Thanks for checking the comments. The patch looks absolutely fine. Maybe it would be a bit too short cut to put into 7.3 on this short term, though it shouldn't be critical. Please keep it on your consideration. Some more refactoring might come yet in 7.4, too.

Thanks.

cmb69 commented:
@weltling Thanks for reviewing!

Unless there are objections, I'm going to merge this into master *after* PHP-7.3 has been branched.

php-pulls commented:
**Comment on behalf of cmb at php.net:**

Applied via c9861bd7a974ef1555395b879c1035a0bbc12ef1.
